### PR TITLE
bitfinex2: Implement some methods

### PR DESCRIPTION
**Pull request by alexeymarunin**
- implement methods: `createOrder`, `cancelOrder`, `cancelAllOrders`, `fetchOrder`,
`fetchOpenOrders`, `fetchClosedOrders`, `fetchStatus`
- implement `handleErrors` method
- add option `fetchOrderOnCreate`

### CODE REVIEW

**Comment by alexeymarunin** [comment]
```
./ccxt/bitfinex2.py:548:14: E999 SyntaxError: invalid syntax
```
I am not quite undestand what I do [wrong](https://travis-ci.org/ccxt/ccxt/builds/636363257#L1838) 🤔

**Comment by kroitor** [comment]
@alexeymarunin there are multiple violations of the codestyle guidelines, explained here:

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#derived-exchange-classes

It would be great if you could clean it up, I'll do my best to merge it asap. Thx!

**Comment by alexeymarunin** [comment]
@kroitor I've fixed all errors)

**Comment by alexeymarunin** [comment]
@kroitor should I do something else for submit merge?

**Comment by kroitor** [comment]
@alexeymarunin it would really help a lot, if you could add the error map and the order status map and change these things:

- https://github.com/ccxt/ccxt/pull/6369/files#diff-0e47289054068ecffe99a953626d4841R894
- https://github.com/ccxt/ccxt/pull/6369/files#diff-0e47289054068ecffe99a953626d4841R707

like this:

- https://github.com/ccxt/ccxt/blob/master/js/bitfinex.js#L1164 
- https://github.com/ccxt/ccxt/blob/master/js/bitfinex.js#L314

------------

... and, if you could also do this:

- https://github.com/ccxt/ccxt/pull/6369/files#diff-0e47289054068ecffe99a953626d4841R577

Like this:

- https://github.com/ccxt/ccxt/blob/master/js/binance.js#L901

------------

... also, if you could remove this call:

- https://github.com/ccxt/ccxt/pull/6369/files#diff-0e47289054068ecffe99a953626d4841R762

and replace it with something like this:

- https://github.com/ccxt/ccxt/blob/master/js/poloniex.js#L1158
- https://github.com/ccxt/ccxt/blob/master/js/poloniex.js#L35

(we cannot do multiple http requests per unified call since that will break the rate-limiting, so the general rule is one request per call)

In other words, now that the syntax is clean, if you could also work on the codestyle – that would really speed up the process. If not, no worries, I will get to it soon.

**Comment by alexeymarunin** [comment]
Okay, I will try )

**Comment by alexeymarunin** [comment]
@kroitor 
* updated `handleErrors()` (add error map)
* updated `fetchOrderTrades()` signature
* added `fetchOrderTradesOnFetchOrder` option to fetch order trades after fetching order. By default, all additional options (`fetchOrderOnCreate` and `fetchOrderTradesOnFetchOrder`) are `false`, so it will take one (max two) requests per call `createOrder` or `fetchOrder`

Unfortunately, order status, in general, is not static string (like in [Binance](https://github.com/ccxt/ccxt/blob/master/js/binance.js#L901)) and can be something like `EXECUTED @ 107.6(-0.2)` (see ORDER_STATUS field in [Order History Request](https://docs.bitfinex.com/reference#rest-auth-orders-history)), so I cannot update `parseOrderStatus`

If I missed something - please, let me know :)

**Comment by kroitor** [comment]
Thank you @alexeymarunin !

> * added `fetchOrderTradesOnFetchOrder` option to fetch order trades after fetching order. By default, all additional options (`fetchOrderOnCreate` and `fetchOrderTradesOnFetchOrder`) are `false`, so it will take one (max two) requests per call `createOrder` or `fetchOrder`

We don't usually do that, since if the user needs this – a fetchOrder by id call can be issued easily on the user side. Otherwise we would add this second request to all exchanges. The rule of thumb, however, is to return just the info provided by the exchange (the id of the order created + some other data if available) and then let the user decide if he needs more details. We would probably add a method called something like `createAndFetch` to the base class that would do the same wrapping, however, at this point we prefer to remain as close to the underlying API as possible.



> Unfortunately, order status, in general, is not static string (like in [Binance](https://github.com/ccxt/ccxt/blob/master/js/binance.js#L901)) and can be something like `EXECUTED @ 107.6(-0.2)` (see ORDER_STATUS field in [Order History Request](https://docs.bitfinex.com/reference#rest-auth-orders-history)), so I cannot update `parseOrderStatus`

In that case we usually add some trivial string parsing, like:

```JavaScript
const statusString = this.safeString (order, 'status');
let status = undefined;
if (statusString !== undefined) {
    const parts = statusString.split (' @ ');
    status = this.parseOrderStatus (this.safeString (parts, 0));
    price = ...
}
```

**Comment by alexeymarunin** [comment]
@kroitor I've updated `fetchOrder`, `createOrder` and `parseOrderStatus` methods.
Please, check my changes

**Comment by npomfret** [comment]
Will you be adding https://docs.bitfinex.com/reference#rest-auth-orders-history ?

**Comment by alexeymarunin** [comment]
@npomfret added in `fetchClosedOrders()`

**Comment by npomfret** [comment]
Excellent.  Looking forward to testing it out.

**Comment by npomfret** [comment]
@alexeymarunin I noticed in the docs that there is an option to omit the symbol in the order history query, so both:

`${url}v2/auth/r/orders/tBTCUSD/hist`
and
`${url}v2/auth/r/orders/hist`,`

will work.


Can you implement both?

**Comment by alexeymarunin** [comment]
@npomfret Done) 
Now it's possible to use `fetchOrder()`, `fetchOpenOrders()` and `fetchClosedOrders()` without `symbol` parameter (fetch for all symbols)

**Comment by alexeymarunin** [comment]
@kroitor 
Is current status of pull request correct?

**Comment by kroitor** [comment]
@alexeymarunin yep, thank you, i am going to review and test it thoroughly and will merge shortly after. Will get to it asap.
